library/x509_crt.c

```diff
@@ -618,7 +618,8 @@ static int x509_get_ext_key_usage( unsigned char **p,
  *      nameAssigner            [0]     DirectoryString OPTIONAL,
  *      partyName               [1]     DirectoryString }
  *
- * NOTE: we only parse and use dNSName at this point.
+ * NOTE: we only parse and use dNSName at this point,
+ * and otherName of type HwModuleName, as defined in RFC 4108.
```

Comment:
I don't think it's in our scope.
I think it's the CA's responsibility to create a proper certificate.
We check whether the extension is critical, and fail if not supported.
Adding logic to check whether an extension exists if the subject distinguished name is
empty is overhead that I think we should not handle

Reply:
> I don't think it's in our scope.
I think it's the CA's responsibility to create a proper certificate.

I think it boils down to "what errors should we expect?". Personally, I would rather check as much as we can, even if the other party is "trusted". There are several instances of issues in the past were a part of a system is compromised and it just breaks everything else because of lack of checking/defensive programming practices. Given that this is a security library there should be even more reason to check.

Reply:
>Given that this is a security library there should be even more reason to check.

I agree with this point, however I still think that it's not part of the scope to add these checks which will probably increase RAM size and complexity. ( adding logci to check empty subject name). It also changes the library's design, as currently don't enforce critical extensions, and we only check that a critical extension is supported

Reply:
Ok, perhaps we should wait for the opinion of other reviewers on what checks we should (or shouldnt) implement. @hanno-arm , what do you think?

Reply:
@hanno-arm @k-stachowiak: Any comments given that you are the other two reviewers?

Reply:
### Security

I don't think this is an immediate security issue, but checking it would promote good security practices:                                                                                                                                          

- We are interpreting and acting on this extension, the criticality poses a 
      security issue if we ignore a critical extension and not when we parse    
      and interpret a non-critical one.                                         
- CAs issuing certificates with SAN set as non-critical extension risk that 
      applications not implementing this extension (for example older versions  
      of Mbed TLS) just ignore it and authenticate whatever identity there is   
      in the Subject DN (empty string probably). This is independent from Mbed  
      TLS and exists whether we check the criticality of this extension or not.
- CAs issuing certificates with non-empty Subject DN, risk ambiguity and
enable reusing certificates and through them keys in different security
contexts, which is a bad practice. But again it is arguably independent
from Mbed TLS and exists whether we check it or not.

### Compliance

I don't think it is a compliance issue, but checking would make us better internet citizens:
                                                                                     
- The standard does not state that applications parsing the certificate     
      must check if it is critical.
- Although not checking it is in line with Postel's robustness principle,
doing so can mask and entrench bugs and have and adverse effect
on the ecosystem and the evolution of the protocol.

### Mbed TLS philosophy

I feel that staying consistent and looking at the big picture we should
still omit these checks.

- X509 is big and complicated, we can't provide a full-featured X509
library, because we want to be very lean and to fit on very constrained
devices.
- Therefore we usually omit many checks that are not critical for security.
Many of them are very similar to these two checks in nature.
- Providing these two checks would not be to much of a burden, but then
we should be consistent in our decisions and implement many other similar
checks. This would mean a significant overhead in code size, a luxury we
can't afford.